src/hnsw/external_index_socket.c

```diff
@@ -108,6 +119,24 @@ static int connect_with_timeout(int sockfd, const struct sockaddr *addr, socklen
     return 0;
 }
 
+int sendall(external_index_socket_t *socket_con, const char *buf, uint32 len, int flags)
+{
+    int total = 0;
+    int bytesleft = len;
+    int n;
+
+    while(total < len) {
+        n = socket_con->send(socket_con, buf + total, bytesleft, flags);
+        if(n == -1 || n == 0) {
+            return n;
```

Comment:
why is this silently returning the number and not closing the socket and erroring out?

it seems we would want to error out from all the callsites anyway. If so, the API is consuming since the caller might not realize they have to check for these values and deal with closing/erroring out since otherwise the function is quite high level (it handles partial sends and retires)